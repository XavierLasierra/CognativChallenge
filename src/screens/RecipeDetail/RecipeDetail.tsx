import React, { useEffect } from "react";
import {
  Text,
  View,
  Image,
  ScrollView,
  SafeAreaView,
  StatusBar,
} from "react-native";
import { IRecipeDetailsProps } from "../../types/interfaces";

import RecipeSummary from "../../components/RecipeSummary/RecipeSummary";
import BackButton from "../../components/BackButton/BackButton";

import placeholder from "../../constants/placeholder.constants";

import styles from "./recipeDetail.styles";
import { AppStyles } from "../../theme";
import { observer } from "mobx-react-lite";
import useStore from "../../hooks/useStore";

const RecipeDetail = observer(
  ({
    navigation,
    route: {
      params: { recipeId },
    },
  }: IRecipeDetailsProps) => {
    const { recipesStore } = useStore();
    const { currentRecipe: recipe } = recipesStore;

    useEffect(() => {
      recipesStore.fetchRecipe(recipeId);
      return () => {
        recipesStore.clearRecipe();
      };
    }, []);

    const handleNavigateBack = () => {
      navigation.pop();
    };

    return (
      <SafeAreaView style={AppStyles.screen.mainScreen}>
        <BackButton actionOnPress={handleNavigateBack} />
        <StatusBar barStyle="dark-content" />
        <View style={AppStyles.container}>
          {recipe && (
            <ScrollView showsVerticalScrollIndicator={false}>
              <View>
                <Image
                  source={{ uri: recipe?.photo || placeholder.defaultImageUrl }}
                  style={styles.recipeImage}
                />
                <View style={styles.darkener} />
              </View>
              <RecipeSummary recipe={recipe} alignCenter />
              <View style={styles.infoBox}>
                <Text style={styles.header}>Ingredients</Text>
                <Text style={styles.description}>{recipe.ingredients}</Text>
              </View>
              <View style={styles.infoBox}>
                <Text style={styles.header}>Info</Text>
                <Text style={styles.description}>{recipe.description}</Text>
              </View>
            </ScrollView>
          )}
        </View>
      </SafeAreaView>
    );
  },
);

export default RecipeDetail;
