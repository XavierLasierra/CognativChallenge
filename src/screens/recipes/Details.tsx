import React from "react";
import {
  Text,
  View,
  Image,
  ScrollView,
  SafeAreaView,
  StatusBar,
} from "react-native";
import { IRecipeDetailsProps } from "../../types/interfaces";

import RecipeSummary from "../../components/RecipeSummary/RecipeSummary";

import placeholder from "../../constants/placeholder.constants.js";

import styles from "./details.styles";
import { AppStyles } from "../../theme";

export default function RecipeDetails({
  route: { params: recipe },
}: IRecipeDetailsProps) {
  return (
    <SafeAreaView style={AppStyles.screen.mainScreen}>
      <StatusBar barStyle="dark-content" />
      <View style={AppStyles.container}>
        <ScrollView showsVerticalScrollIndicator={false}>
          <View>
            <Image
              source={{ uri: recipe?.photo || placeholder.defaultImageUrl }}
              style={styles.recipeImage}
            />
            <View style={styles.darkener} />
          </View>
          <RecipeSummary recipe={recipe} alignCenter />
          <View style={styles.infoBox}>
            <Text style={styles.header}>Ingredients</Text>
            <Text style={styles.description}>{recipe.ingredients}</Text>
          </View>
          <View style={styles.infoBox}>
            <Text style={styles.header}>Info</Text>
            <Text style={styles.description}>{recipe.description}</Text>
          </View>
        </ScrollView>
      </View>
    </SafeAreaView>
  );
}
