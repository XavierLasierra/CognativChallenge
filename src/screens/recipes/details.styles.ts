import { StyleSheet } from "react-native";

import Colors from "../../theme/Colors";
import Metrics from "../../theme/Metrics";

export default StyleSheet.create({
  darkener: {
    backgroundColor: Colors.black30,
    bottom: 0,
    left: 0,
    position: "absolute",
    right: 0,
    top: 0,
    zIndex: 5,
  },
  description: {
    color: Colors.darkGrey,
    fontSize: 14,
    lineHeight: 24,
  },
  header: {
    color: Colors.black,
    fontSize: 14,
    lineHeight: 24,
    textAlign: "center",
  },
  infoBox: {
    alignItems: "stretch",
    alignSelf: "stretch",
    justifyContent: "center",
    paddingHorizontal: Metrics.baseSpace,
    paddingVertical: Metrics.smallSpace,
  },
  recipeImage: {
    height: 192,
  },
});
