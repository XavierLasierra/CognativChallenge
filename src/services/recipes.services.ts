import axios from "axios";
import { API_BASE_URL } from "@env";
import {
  IAxiosRecipeData,
  IAxiosRecipesData,
  IRecipe,
} from "../types/interfaces";

export async function getRecipes(): Promise<IRecipe[]> {
  try {
    const { data }: IAxiosRecipesData = await axios.get(
      `${API_BASE_URL}/recipes`,
    );

    const { data: recipes } = data;
    return recipes || [];
  } catch (error) {
    return [];
  }
}

export async function getRecommendedRecipes(): Promise<IRecipe[]> {
  try {
    const { data }: IAxiosRecipesData = await axios.get(
      `${API_BASE_URL}/recipes?recommended=1`,
    );

    const { data: recommendedRecipes } = data;
    return recommendedRecipes || [];
  } catch (error) {
    return [];
  }
}

export async function getRecipe(recipeId: string): Promise<IRecipe | null> {
  try {
    const { data: recipe }: IAxiosRecipeData = await axios.get(
      `${API_BASE_URL}/recipes/${recipeId}`,
    );
    return recipe || null;
  } catch (error) {
    return null;
  }
}
