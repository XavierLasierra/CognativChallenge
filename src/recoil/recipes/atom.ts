import { atom, selector } from "recoil";
import { getRecipes } from "../../services/recipes.services";
import { IRecipe } from "../../types/interfaces";

export default atom<IRecipe[]>({
  key: "Recipes",
  default: selector({
    key: "Recipes/Default",
    get: () => getRecipes(),
  }),
});
