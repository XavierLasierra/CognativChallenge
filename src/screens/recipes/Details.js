import React, { Component } from "react";
import {
  Text,
  StyleSheet,
  View,
  Image,
  ScrollView,
  SafeAreaView,
  StatusBar,
} from "react-native";

import Colors from "../../theme/Colors";
import Metrics from "../../theme/Metrics";

import recipes from "../../__mocks__/recipes.mock";

const [data] = recipes;

export default class RecipeDetails extends Component {
  render() {
    let imageUrl = "http://via.placeholder.com/640x360";
    if (data && data.photo) {
      imageUrl = data.photo;
    }

    return (
      <SafeAreaView style={styles.mainScreen}>
        <StatusBar barStyle="dark-content" />
        <View style={styles.container}>
          <ScrollView showsVerticalScrollIndicator={false}>
            <View style={styles.imageContainer}>
              <Image source={{ uri: imageUrl }} style={styles.recipeImage} />
              <View style={styles.darkener} />
            </View>
            <View style={styles.infoContainer}>
              <Text style={styles.category}>{data.categoryName}</Text>
              <Text style={styles.title}>{data.name}</Text>
              <View style={styles.properties}>
                <View style={styles.cell}>
                  <Text style={styles.cellText}>{data.duration} minutes</Text>
                </View>
                <View style={styles.cell}>
                  <Text style={styles.cellText}>{data.complexity}</Text>
                </View>
                <View style={styles.cell}>
                  <Text style={styles.cellText}>{data.people} people</Text>
                </View>
              </View>
            </View>
            <View style={styles.infoBox}>
              <Text style={styles.header}>Ingredients</Text>
              <Text style={styles.description}>{data.ingredients}</Text>
            </View>
            <View style={styles.infoBox}>
              <Text style={styles.header}>Info</Text>
              <Text style={styles.description}>{data.description}</Text>
            </View>
          </ScrollView>
        </View>
      </SafeAreaView>
    );
  }
}

const styles = StyleSheet.create({
  category: {
    color: Colors.darkGrey,
    fontSize: 10,
    fontStyle: "normal",
    fontWeight: "bold",
    letterSpacing: 1,
    lineHeight: 24,
    textAlign: "center",
  },
  cell: {
    alignItems: "center",
    flex: 1,
    flexDirection: "row",
    justifyContent: "center",
    marginRight: Metrics.smallSpace,
  },
  cellText: {
    color: Colors.mediumGrey,
    fontSize: 13,
    fontStyle: "normal",
    fontWeight: "300",
    letterSpacing: 0.5,
    lineHeight: 24,
  },
  container: {
    alignItems: "stretch",
    backgroundColor: Colors.mainBackground,
    flex: 1,
    justifyContent: "flex-start",
  },
  darkener: {
    backgroundColor: Colors.black30,
    bottom: 0,
    left: 0,
    position: "absolute",
    right: 0,
    top: 0,
    zIndex: 5,
  },
  description: {
    color: Colors.darkGrey,
    fontSize: 14,
    lineHeight: 24,
  },
  header: {
    color: Colors.black,
    fontSize: 14,
    lineHeight: 24,
    textAlign: "center",
  },
  imageContainer: {},
  infoBox: {
    alignItems: "stretch",
    alignSelf: "stretch",
    justifyContent: "center",
    paddingHorizontal: Metrics.baseSpace,
    paddingVertical: Metrics.smallSpace,
  },
  mainScreen: {
    backgroundColor: Colors.mainBackground,
    flex: 1,
  },
  properties: {
    alignItems: "center",
    flexDirection: "row",
    justifyContent: "flex-start",
  },
  recipeImage: {
    height: 192,
  },
  title: {
    fontWeight: "bold",
  },
  title: {
    color: Colors.black,
    fontSize: 16,
    fontStyle: "normal",
    fontWeight: "500",
    letterSpacing: 0.5,
    lineHeight: 24,
    textAlign: "center",
  },
});
