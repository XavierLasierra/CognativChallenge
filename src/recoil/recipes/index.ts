import atom from "./atom";
import recipeById from "./recipeById";

export { recipeById };
export default atom;
