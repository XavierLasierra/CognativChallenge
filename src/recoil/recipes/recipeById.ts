import { selectorFamily } from "recoil";
import recipes from "./atom";

export default selectorFamily({
  key: "RecipeById",
  get:
    recipeId =>
    ({ get }) => {
      return get(recipes).find(({ _id }) => _id === recipeId);
    },
});
