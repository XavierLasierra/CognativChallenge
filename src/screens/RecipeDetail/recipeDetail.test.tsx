import React from "react";
import RecipeDetail from "./RecipeDetail";
import { render } from "../../utils/test.utils";

import { getRecipes } from "../../services/recipes.services";
import recipesMock from "../../__mocks__/recipes.mock";

jest.mock("../../services/recipes.services");

describe("Given a RecipeDetails component", () => {
  describe("When it is rendered", () => {
    test("Then should match the snapshot", () => {
      (getRecipes as jest.Mock).mockReturnValue(recipesMock);
      const navigation = {
        pop: jest.fn(),
        push: jest.fn(),
      };
      const route = {
        params: {
          recipeId: "52944",
        },
      };

      const screen = render(
        <RecipeDetail navigation={navigation} route={route} />,
      );

      expect(screen).toMatchSnapshot();
    });
  });
});
