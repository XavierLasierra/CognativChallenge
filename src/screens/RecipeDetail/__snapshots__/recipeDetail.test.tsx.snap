// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a RecipeDetails component When it is rendered Then should match the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#f7f7f7",
      "flex": 1,
    }
  }
>
  <View
    accessibilityRole="button"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 50,
        "left": 20,
        "opacity": 1,
        "position": "absolute",
        "top": 30,
        "width": 50,
        "zIndex": 2,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#e0e0e0",
            "borderRadius": 50,
            "height": 7,
            "position": "absolute",
            "width": 25,
          },
          Object {
            "transform": Array [
              Object {
                "rotate": "-45deg",
              },
            ],
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#e0e0e0",
            "borderRadius": 50,
            "height": 7,
            "position": "absolute",
            "width": 25,
          },
          Object {
            "top": 13,
            "transform": Array [
              Object {
                "rotate": "45deg",
              },
            ],
          },
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "backgroundColor": "#f7f7f7",
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <RCTScrollView
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View>
          <Image
            source={
              Object {
                "uri": "https://www.themealdb.com/images/media/meals/1520084413.jpg",
              }
            }
            style={
              Object {
                "height": 192,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "rgba(0, 0, 0, 0.3)",
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 1,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "stretch",
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#9e9e9e",
                "fontSize": 10,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "letterSpacing": 1,
                "lineHeight": 24,
                "textAlign": "center",
              }
            }
          >
            Fish
          </Text>
          <Text
            style={
              Object {
                "color": "#333333",
                "fontSize": 16,
                "fontStyle": "normal",
                "fontWeight": "500",
                "letterSpacing": 0.5,
                "lineHeight": 24,
                "textAlign": "center",
              }
            }
          >
            Escovitch Fish
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "flex-start",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginRight": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#bdbdbd",
                    "fontSize": 13,
                    "fontStyle": "normal",
                    "fontWeight": "300",
                    "letterSpacing": 0.5,
                    "lineHeight": 24,
                  }
                }
              >
                11
                 minutes
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginRight": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#bdbdbd",
                    "fontSize": 13,
                    "fontStyle": "normal",
                    "fontWeight": "300",
                    "letterSpacing": 0.5,
                    "lineHeight": 24,
                  }
                }
              >
                Hard
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginRight": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#bdbdbd",
                    "fontSize": 13,
                    "fontStyle": "normal",
                    "fontWeight": "300",
                    "letterSpacing": 0.5,
                    "lineHeight": 24,
                  }
                }
              >
                3
                 people
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "stretch",
              "alignSelf": "stretch",
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#333333",
                "fontSize": 14,
                "lineHeight": 24,
                "textAlign": "center",
              }
            }
          >
            Ingredients
          </Text>
          <Text
            style={
              Object {
                "color": "#9e9e9e",
                "fontSize": 14,
                "lineHeight": 24,
              }
            }
          >
            2 eggs
4 tomatoes
salt
pepper
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "stretch",
              "alignSelf": "stretch",
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#333333",
                "fontSize": 14,
                "lineHeight": 24,
                "textAlign": "center",
              }
            }
          >
            Info
          </Text>
          <Text
            style={
              Object {
                "color": "#9e9e9e",
                "fontSize": 14,
                "lineHeight": 24,
              }
            }
          >
            Rinse fish; rub with lemon or lime, seasoned with salt and pepper or use your favorite seasoning. I used creole seasoning. Set aside or place in the oven to keep it warm until sauce is ready.

In large skillet heat oil over medium heat, until hot, add the fish, cook each side- for about 5-7 minutes until cooked through and crispy on both sides. Remove fish and set aside. Drain oil and leave about 2-3 tablespoons of oil
Add, bay leave, garlic and ginger, stir-fry for about a minute making sure the garlic does not burn
Then add onion, bell peppers, thyme, scotch bonnet, sugar, all spice-continue stirring for about 2-3 minutes. Add vinegar, mix an adjust salt and pepper according to preference. Let it simmer for about 2 more minutes. 

Discard bay leave, thyme spring and serve over fish with a side of this bammy. You may make the sauce about 2 days in advance.
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
